mvpa2/measures/label.py

```diff
@@ -0,0 +1,139 @@
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the PyMVPA package for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Label adjacent features in a dataset
+"""
+import numpy as np
+from collections import defaultdict
+from operator import itemgetter
+
+from ..base.param import Parameter
+from ..base.constraints import EnsureBool
+from ..datasets.base import Dataset
+from ..misc.neighborhood import CachedQueryEngine
+
+from .base import Measure
+
+if __debug__:
+    from ..base import debug
+
+
+class ClustersLabeler(Measure):
+    """Determines contiguous regions of features using a query engine.
+
+    "label" in the name chosen to mimic scipy.ndimage.measurements.label
+    """
+    qe = Parameter(
+        None,
+        doc="QueryEngine to determine neighbors of any given feature")
```

Comment:
On Fri, 08 Apr 2016, bpinsard wrote:

>    In mvpa2/measures/label.py:

>  > +from ..misc.neighborhood import CachedQueryEngine
>  > +
>  > +from .base import Measure
>  > +
>  > +if __debug__:
>  > +    from ..base import debug
>  > +
>  > +
>  > +class ClustersLabeler(Measure):
>  > +    """Determines contiguous regions of features using a query engine.
>  > +
>  > +    "label" in the name chosen to mimic scipy.ndimage.measurements.label
>  > +    """
>  > +    qe = Parameter(
>  > +        None,
>  > +        doc="QueryEngine to determine neighbors of any given feature")

>    Will the cluster computation depend on the "radius" of the query engine?

if query engine takes radius, then yes ;)  but the point is to allow any
neighborhood querying by providing corresponding query engine.  E.g. in
the volume it could be the IndexQueryEngine which would take some
HollowSphere in one dimension (e.g. voxels coords) and then some other,
e.g. Sphere in another (e.g. time) thus allowing even to cluster
spatio-temporal beasts of arbitrary kinds ;)

>    Or would it be possible to extract a neighborhood structure that takes
>    only the closest neighboring samples?

define "samples" ;)  those query engines operate across "features" (in
pymvpa terminology), so if you need to operate across samples, you would
need to bring them into the same sample (e.g. boxcar'ing etc)